enhancements/kube-apiserver/audit-policy.md

```diff
@@ -261,61 +284,302 @@ the respective file-based audit policy of
+    // +required
+    Group string `json:"group"`
+
+    // profile specifies the name of the desired audit policy configuration to be deployed to
+    // all OpenShift-provided API servers in the cluster.
+    //
+    // The following profiles are provided:
+    // - Default: the existing default policy.
+    // - WriteRequestBodies: like 'Default', but logs request and response HTTP payloads for
+    // write requests (create, update, patch).
+    // - AllRequestBodies: like 'WriteRequestBodies', but also logs request and response
+    // HTTP payloads for read requests (get, list).
+    // - None: no requests are logged at all, not even oauthaccesstokens and oauthauthorizetokens.
+    //
+    // If unset, the 'Default' profile is used as the default.
+    //
+    // +kubebuilder:Enum=Default,WriteRequestBodies,AllRequestBodies,None
+    // +kubebuilder:validation:Required
+    // +required
+    Profile AuditProfileType `json:"profile,omitempty"`
+}
+```
+
+### Library-go Implementation of CustomRules
+
+The library-go interface for audit policies before addition of `customRules` in `github.com/openshift/library-go/pkg/operator/apiserver/audit` looked like this:
+
+```go
+type AuditPolicyPathGetterFunc func(profile string) (string, error)
+
+func DefaultPolicy() ([]byte, error)
+func WithAuditPolicies(targetName string, targetNamespace string, assetDelegateFunc resourceapply.AssetFunc) resourceapply.AssetFunc
+func GetAuditPolicies(targetName, targetNamespace string) (*corev1.ConfigMap, error)
+func NewAuditPolicyPathGetter(path string) (AuditPolicyPathGetterFunc, error)
+```
+
+with a config observer in `github.com/openshift/library-go/pkg/operator/configobserver/apiserver`:
+
+```go
+func NewAuditObserver(pathGetter AuditPolicyPathGetterFunc) configobserver.ObserveConfigFunc
+```
+
+The asset func is wired into the static resource controller in our operators like this:
+
+```go
+apiservercontrollerset.NewAPIServerControllerSet(
+    ...
+).WithStaticResourcesController(
+    "APIServerStaticResources",
+    libgoassets.WithAuditPolicies("audit", operatorclient.TargetNamespace, v311_00_assets.Asset),
+    []string{
+	    ...,
+		libgoassets.AuditPoliciesConfigMapFileName,
+	},
+	...,
+).WithRevisionController(
+	operatorclient.TargetNamespace,
+	[]revision.RevisionResource{
+		{
+			Name: "audit",
+		},
+	},
+	...,
+)
+```
+
+As the audit policy file is not static anymore with the addition of
+custom rules, this approach will not work anymore. Instead, we need a
+dynamic audit policy controller in the apiserver operators that
+
+1. computes the policy depending on the apiserver configuration
+   resource
+2. and copies that as a `ConfigMap` into the operand namespace.
+
+Then the revision controller will notice it changing and assigns a new
+operand revision that causes a rollout. We will add the audit policy
+controller to the APIServer controller set in library-go, shared by
+our three apiserver operators:
```

Comment:
When you say audit policy controller here, is this a reference to a new function that should be created?

Reply:
yes, a new, small controller in the apiserver operator set in library-go. It will continuously compute the necessary audit policy and writes it to the ConfigMap when it changes.